helper/schema/resource.go

```diff
@@ -348,31 +368,31 @@ func (r *Resource) Apply(
 		// If we're only destroying, and not creating, then return
 		// now since we're done!
 		if !d.RequiresNew() {
-			return nil, nil
+			return nil, diags
 		}
 
 		// Reset the data to be stateless since we just destroyed
 		data, err = schemaMap(r.Schema).Data(nil, d)
-		// data was reset, need to re-apply the parsed timeouts
-		data.timeouts = &rt
 		if err != nil {
-			return nil, err
+			return nil, append(diags, diag.FromErr(err))
 		}
+
+		// data was reset, need to re-apply the parsed timeouts
+		data.timeouts = &rt
 	}
 
-	err = nil
 	if data.Id() == "" {
 		// We're creating, it is a new resource.
 		data.MarkNewResource()
-		err = r.create(ctx, data, meta)
+		diags = append(diags, r.create(ctx, data, meta)...)
 	} else {
 		if !r.updateFuncSet() {
-			return s, fmt.Errorf("doesn't support update")
+			return s, append(diags, diag.FromErr(fmt.Errorf("doesn't support update")))
```

Comment:
not a big deal, but if this isn't actually formatting, can you just construct the struct manually here? the implementation of `Errorf` is non-trivial at least and pretty unnecessary here: https://golang.org/src/fmt/errors.go?s=624:674#L7